src/compiler/program.ts

```diff
@@ -1454,7 +1569,79 @@ namespace ts {
             }
         }
 
+        /**
+         * ExtensionKind.SyntacticLint or ExtensionKind.SemanticLint only
+         */
+        function performLintPassOnFile(sourceFile: SourceFile, kind: ExtensionKind): Diagnostic[] | undefined {
+            const lints = compilerExtensions[kind];
+            if (!lints || !lints.length) {
+                return;
+            }
+            type UniqueLint = {name: string, walker: LintWalker, accepted: boolean};
+            const initializedLints = new Array<UniqueLint>(lints.length);
+            const diagnostics: Diagnostic[] = [];
+            let activeLint: UniqueLint;
+            let parent: Node | undefined = undefined;
+            for (let i = 0; i < lints.length; i++) {
+                if (kind === ExtensionKind.SemanticLint) {
+                    initializedLints[i] = {name: lints[i].name, walker: new (lints[i].ctor as SemanticLintProviderStatic)(ts, getTypeChecker(), lints[i].args), accepted: true};
+                }
+                else if (kind === ExtensionKind.SyntacticLint) {
+                    initializedLints[i] = {name: lints[i].name, walker: new (lints[i].ctor as SyntacticLintProviderStatic)(ts, lints[i].args), accepted: true};
+                }
+            }
+
+            visitNode(sourceFile);
+
+            return diagnostics;
+
+            function visitNode(node: Node) {
+                let oneAccepted = false;
+                const oldParent = parent;
+                const needsReset: boolean[] = new Array(initializedLints.length);
+                for (let i = 0; i < initializedLints.length; i++) {
+                    if (initializedLints[i].accepted) {
+                        activeLint = initializedLints[i];
+                        activeLint.accepted = false;
+                        node.parent = parent;
+                        activeLint.walker.visit(node, accept, error);
+                        if (activeLint.accepted) {
+                            oneAccepted = true;
+                        }
+                        else {
+                            needsReset[i] = true;
+                        }
+                    }
+                }
+                parent = node;
+                if (oneAccepted) {
+                    forEachChild(node, visitNode);

```

Comment:
For performance reasons, I see how this makes sense, but this might create situations where users have to correct multiple passes of errors. For example, a rule banning `while` loops would only handle the outermost case of nested `while` loops